src/goto-programs/remove_instanceof.cpp

```diff
@@ -0,0 +1,255 @@
+    lowered_count(0)
+  {
+    class_hierarchy(_symbol_table);
+  }
+
+  // Lower instanceof for all functions:
+  void lower_instanceof();
+
+protected:
+  symbol_tablet &symbol_table;
+  namespacet ns;
+  class_hierarchyt class_hierarchy;
+  goto_functionst &goto_functions;
+  int lowered_count;
+
+  bool lower_instanceof(goto_programt &);
+
+  typedef std::vector<
+    std::pair<goto_programt::targett, goto_programt::targett>> instanceof_instt;
+
+  void lower_instanceof(
+    goto_programt &,
+    goto_programt::targett,
+    instanceof_instt &);
+
+  void lower_instanceof(
+    exprt &,
+    goto_programt &,
+    goto_programt::targett,
+    instanceof_instt &);
+};
+
+/*******************************************************************\
+
+Function: remove_instanceoft::lower_instanceof
+
+  Inputs: Expression to lower `expr` and the `goto_program` and
+          instruction `this_inst` it belongs to.
+
+ Outputs: Side-effect on `expr` replacing it with an explicit clsid test
+
+ Purpose: Replaces an expression like
+          e instanceof A
+          with
+          e.@class_identifier == "A"
+          Or a big-or of similar expressions if we know of subtypes
+          that also satisfy the given test.
+
+\*******************************************************************/
+
+void remove_instanceoft::lower_instanceof(
+  exprt &expr,
+  goto_programt &goto_program,
+  goto_programt::targett this_inst,
+  instanceof_instt &inst_switch)
+{
+  if(expr.id()==ID_java_instanceof)
+  {
+    const exprt &check_ptr=expr.op0();
+    assert(check_ptr.type().id()==ID_pointer);
+    const exprt &target_arg=expr.op1();
+    assert(target_arg.id()==ID_type);
+    const typet &target_type=target_arg.type();
+
+    // Find all types we know about that satisfy the given requirement:
+    assert(target_type.id()==ID_symbol);
+    const irep_idt &target_name=
+      to_symbol_type(target_type).get_identifier();
+    std::vector<irep_idt> children=
+      class_hierarchy.get_children_trans(target_name);
+    children.push_back(target_name);
+
+    assert(!children.empty() && "Unable to execute instanceof");
+
+    // Insert an instruction before this one that assigns the clsid we're
+    // checking against to a temporary, as GOTO program if-expressions should
+    // not contain derefs.
+
+    symbol_typet jlo("java::java.lang.Object");
+    exprt object_clsid=get_class_identifier_field(check_ptr, jlo, ns);
+
+    std::ostringstream symname;
+    symname << "instanceof_tmp::instanceof_tmp" << (++lowered_count);
+    auxiliary_symbolt newsym;
+    newsym.name=symname.str();
+    newsym.type=object_clsid.type();
+    newsym.base_name=newsym.name;
+    newsym.mode=ID_java;
+    newsym.is_type=false;
+    assert(!symbol_table.add(newsym));
+
+    code_assignt clsid_tmp(newsym.symbol_expr(), object_clsid);
+    auto newinst=goto_program.insert_after(this_inst);
+    newinst->make_assignment();
+    newinst->code=std::move(clsid_tmp);
```

Comment:
I'll keep asking the "why std::move?" questions until I receive an "It's better, because ..." answer ;-)